### PR TITLE
[WIP]Add a new helper method for evalf

### LINKED ISSUE #20479: No proper way to recurse in evalf
The problem is that in the FiniteSet case evalf recurses to the args like this:
https://github.com/sympy/sympy/blob/b3cae120dad022dd83e338df69c45f60dd95efd9/sympy/sets/sets.py#L1886-L1887

Since that calls `_evalf` the call will go through to `_eval_evalf` which in the case of an `Add` means `AssocOp._eval_evalf`:
https://github.com/sympy/sympy/blob/b3cae120dad022dd83e338df69c45f60dd95efd9/sympy/core/operations.py#L354

However calling `x.evalf()` leads to `evalf(x, ...)` where the `evalf` function is the internal function here:
https://github.com/sympy/sympy/blob/b3cae120dad022dd83e338df69c45f60dd95efd9/sympy/core/evalf.py#L1321

The dispatches through the table to call `evalf_add`:
https://github.com/sympy/sympy/blob/b3cae120dad022dd83e338df69c45f60dd95efd9/sympy/core/evalf.py#L519

The `evalf_add` function has a special handler to ensure accuracy in the face of cancellation.

It seems that when recursing down the args in an evalf call the function to use should be `evalf(x, ...)` (as is done in the internal routines in the evalf module). The alternatives are all problematic:
1) The `x.evalf()` method takes the wrong kind of precision and is the public method.
2) The `x._evalf()` method just called `x._eval_evalf` and handles a `None` return. Perhaps this method isn't needed at all.
3) The `x._eval_evalf` method bypasses the table so it won't e.g. lead to `evalf_add`.

However the `evalf(x, ...)` function also isn't suitable for recursion from `x._eval_evalf` because the `_eval_evalf` method doesn't receive the arguments that should be passed to `evalf(x, ...)`. The `_eval_evalf` method is clearly the way for non-core classes to implement `evalf` but there does not seem to be a proper way for such a method to recurse to its args.

This all seems like a bit of a mess to me and indeed usage around the codebase of `.evalf`, `._evalf` and `._eval_evalf` is inconsistent.

_Originally posted by @oscarbenjamin in https://github.com/sympy/sympy/issues/20379#issuecomment-730563821_

### PR DESCRIPTION
**Pull request by sidhu1012**
#### References to other Issues or PRs

Fixes #20479 

#### Brief description of what is fixed or changed
Added a new helper method for `evalf` to make `evalf` consistent without compromising `precision` and `options` parameter.

#### Other comments
Read discussion #20479 

#### Release Notes

* core
    * Added a new helper method for evalf()

### CODE REVIEW

File: sympy/core/relational.py [hunk 1]
```diff
@@ -196,8 +196,17 @@ def negated(self):
...
+        #from sympy.core.evalf import evalf
+        subs, quad = None, None
+        if 'subs' in options:
+            subs = options['subs']
+        if 'quad' in options:
+            quad = options['quad']
+        #return self.func(*[evalf(s, prec, options) for s in self.args])
+        return self.func(*[s.evalf(n, subs=subs, maxn=options['maxn'], chop=options['chop'],
+                            strict=options['strict'], quad=quad,
+                            verbose=options['verbose']) for s in self.args])
```
Comment (oscarbenjamin) [inline sympy/core/relational.py:209]: This is a horrible way to recurse. We should make it so that `_eval_evalf_options` takes the same arguments as the `evalf` function and then recurse using the `evalf` function.

**Comment by czgdp1807** [comment]
Do we need a release notes entry for this?

File: sympy/core/relational.py (hunk 1 above)
Comment (sidhu1012) [reply sympy/core/relational.py]: Yes, agreed.

**Comment by sidhu1012** [comment]
@oscarbenjamin I broke `.evalf` function in 2 parts, one part is the public `.evalf` function which handles the preprocessing and calls the other part. The other part calls the `evalf` from `sympy.core.evalf` and handles some corner cases which are required in the codebase.

**Comment by oscarbenjamin** [comment]
This looks like the right idea but there needs to be comments and docs explaining how to use this. We also need to make sure that this works with all the different combinations of how downstream libraries and users might be depending on the existing behaviour. There hasn't been clear documentation for users so there can be many combinations of how they might have implemented these things (there is inconsistency in the sympy codebase so I expect inconsistency elsewhere as well).

Somewhere the documentation should explain how Basic subclasses should implement their own evalf routines probably in the class docstring for `EvalfMixin` which is currently very short:
https://docs.sympy.org/latest/modules/core.html#sympy.core.evalf.EvalfMixin

I think a docs example implementing a simple `Basic` subclass with `_eval_evalf_options` would be useful. This should clarify that a subclass should inherit `EvalfMixin` and should override the `_eval_evalf_options` method and should recurse using `._evalf_options`. There should also be notes explaining that the `_eval_evalf` and `_evalf` methods are retained/used only for backwards compatibility.

There should also be some reference to `EvalfMixin` (perhaps a see also) in the `evalf` docstring referring to there for how to implement evalf behaviour in user-created classes.

Anywhere in the code where `_eval_evalf_options` or `_eval` or `_eval_evalf` is used should have comments discussing why the different methods are being called (e.g. for backwards compatibility). Each of these methods should have a docstring indicating what it is to be used for and whether it is recommended or deprecated.

To ensure backwards compatibility we need to test the behaviour of different possible subclasses that users might have made in the past. A user might have made a class which is

1. An Expr subclass
2. A Basic subclass
3. A Basic subclass that also subclasses EvalfMixin

That subclass might implement:

1. `._eval_evalf`
2. `._eval`
3. `.evalf`
4. Some combination of the above
4. None of the above

That gives a lot of combinations for how someone might have implemented this. We should test that all combinations still give the same results after this change. We should test the expected behaviour for precision, subs etc.

We also need to test in reverse. Some downstream user code might be calling methods like `_eval_evalf` or `_eval` on standard sympy objects both Basic and Expr. We need to test that their code will still work correctly (or at least the same as before) when given a class that implements `_eval_evalf_options` instead.

Also a more extensive release note is needed. We need to make sure that anyone who has implemented an `_eval_evalf` method knows that they should add an `_eval_evalf_options` method as well. They might need to add both methods in order to retain compatibility with different sympy versions (so we need to test that we get the expected behaviour when *both* methods are defined). Also if they have used `.evalf` or `._evalf` or `._eval_evalf` to recurse then they also need to change their code and there needs to be an explanation of how to change it while maintaining compatibility with older sympy versions (and we need to test that their changed code works with the new sympy code).

We should clearly indicate that the use of either `_eval_evalf` or `_evalf` is deprecated although I don't think that `DeprecationWarning` should be used. Instead the docs can use the `.. deprecated::` directive and the code should have comments clearly indicating where deprecated code is being called for backwards compatibility.

I think it probably makes sense to have a function `evalf_options` in the `evalf` module whose purpose is to check for the different methods `._eval_evalf_options`, `._eval_evalf`, '._eval` and call the appropriate implementation. Then we can use that in the evalf module any time we need to call one of these methods. Potentially that is also helpful for users who want to support the new sympy code as well as the old because we can suggest for them to do something like (not sure if this actually works in all cases):
```python
try:
    # sympy version >= 1.8:
    from sympy.core.evalf import evalf_options
except ImportError:
    # sympy version < 1.8
    def evalf_options(obj, prec, options):
        return obj._eval_evalf(prec)

class MyClass(Expr):
    def _eval_evalf_options(self, prec, options):
        args = [evalf_options(a, prec, options) for a in self.args]
        # new code that makes use of options

    def _eval_evalf(self, prec):
        # old version that doesn't make use of options
```
Basically it is not enough to ensure that the sympy codebase is internally consistent because I think this will be used a lot by downstream code. We need to make sure that things that used to work continue to work. We also need to document for users how they should ideally implement things and how to make their code work with both old and new version of sympy. We also need tests for everything.

**Comment by sidhu1012** [comment]
@oscarbenjamin I will update the docstring wherever necessary.


> That gives a lot of combinations for how someone might have implemented this. We should test that all combinations still give the same results after this change. We should test the expected behaviour for precision, subs etc.
> 
> We also need to test in reverse. Some downstream user code might be calling methods like `_eval_evalf` or `_eval` on standard sympy objects both Basic and Expr. We need to test that their code will still work correctly (or at least the same as before) when given a class that implements `_eval_evalf_options` instead.
> 

`_evalf_options` is a private function which is called by public `.evalf` so that doesn't need and changes downstream , the only change required is to replace `_eval_evalf` with `eval_evalf_options` so that `options` parameters are considered and recurse on `evalf_options` to skip preprocessing.
The changes that I have done would first look for `_eval_evalf_options` and if it's not defined(gives `None` or class is using `_eval_evalf`) then it would revert back to `eval_evalf`, thus won't break backward compatibility but would just prefer `_eval_evalf_options` over `_eval_evalf` . `_eval` just calls  `_eval_evalf` so that won't be an isse.

So I think all possible combinations of subclasses using `evalf` have been covered.



> I think it probably makes sense to have a function `evalf_options` in the `evalf` module whose purpose is to check for the different methods `._eval_evalf_options`, `._eval_evalf`, '._eval` and call the appropriate implementation. Then we can use that in the evalf module any time we need to call one of these methods. Potentially that is also helpful for users who want to support the new sympy code as well as the old because we can suggest for them to do something like (not sure if this actually works in all cases):

I don't think it's a good idea because `._evalf_options` and `evalf()` would take care of implemented methods and would call them appropriately .

**Comment by sidhu1012** [comment]
@oscarbenjamin How to define classes in docstring?

**Comment by oscarbenjamin** [comment]
> So I think all possible combinations of subclasses using `evalf` have been covered.

Can you add tests for all cases?

**Comment by sidhu1012** [comment]
> Can you add tests for all cases?

Yes I would in next commits.

**Comment by sidhu1012** [comment]
@oscarbenjamin I have added test cases for all possible combinations of EvalfMixin. Please review.

**Comment by sidhu1012** [comment]
I replaced `eval_evalf` with `eval_evalf_options` randomly at many places and is working fine and isn't breaking code plus it also corrected precision at many places and all `options` are working fine. Example `subs` with `FiniteSet`.

**Comment by sidhu1012** [comment]
@oscarbenjamin Ready for review!

**Comment by oscarbenjamin** [comment]
These tests are quite repetitive. You should combine them into a single function and make a superclass to absorb the repeated code.

You are testing lots of combinations of classes that all only inherit from `EvalfMixin`. That's one possible case but there are others such as a subclass of Expr (which already subclasses EvalfMixin) and a subclass of Basic that does subclass EvalfMixin and also one that does not.

Crucially we need to think about cases where someone has not subclassed EvalfMixin at all and has just implemented a method like `_eval_evalf`. In current master I can do this:
```
In [70]: class A(Basic): 
    ...:     def _evalf(self, prec): 
    ...:         return Float('2.0', prec) 
    ...:                                                                                                                          

In [71]: FiniteSet(A()).evalf()                                                                                                   
Out[71]: {2.0}
```
With this PR that gives:
```
AttributeError: 'A' object has no attribute 'evalf'
```

**Comment by sidhu1012** [comment]
> Crucially we need to think about cases where someone has not subclassed EvalfMixin at all and has just implemented a method like `_eval_evalf`. In current master I can do this:
> 
> ```
> In [70]: class A(Basic): 
>     ...:     def _evalf(self, prec): 
>     ...:         return Float('2.0', prec) 
>     ...:                                                                                                                          
> 
> In [71]: FiniteSet(A()).evalf()                                                                                                   
> Out[71]: {2.0}
> ```
> 
> With this PR that gives:
> 
> ```
> AttributeError: 'A' object has no attribute 'evalf'
> ```

In current master it gives `AttributeError: 'A' object has no attribute 'evalf'` and so does in this PR.

**Comment by sidhu1012** [comment]
> > Crucially we need to think about cases where someone has not subclassed EvalfMixin at all and has just implemented a method like `_eval_evalf`. In current master I can do this:
> > ```
> > In [70]: class A(Basic): 
> >     ...:     def _evalf(self, prec): 
> >     ...:         return Float('2.0', prec) 
> >     ...:                                                                                                                          
> > 
> > In [71]: FiniteSet(A()).evalf()                                                                                                   
> > Out[71]: {2.0}
> > ```
> > 
> > 
> > With this PR that gives:
> > ```
> > AttributeError: 'A' object has no attribute 'evalf'
> > ```
> 

In current master it gives `AttributeError: 'A' object has no attribute 'evalf'` and so does in this PR.

**Comment by sidhu1012** [comment]
@oscarbenjamin Since the code you showed is giving the same `AttributeError` should I make it work in this PR ?

**Comment by sidhu1012** [comment]
@oscarbenjamin Since both the master branch and this PR both are giving `AttributeError`  what shall I do?

**Comment by oscarbenjamin** [comment]
> In current master it gives `AttributeError: 'A' object has no attribute 'evalf'` and so does in this PR.

Sorry, I should have said that in sympy 1.7 it works. It seems to have already broken on master from another evalf PR.

Maybe that particular case isn't so important but we do need to be careful to make sure that anything that worked in sympy 1.6 or 1.7 still works if possible.

**Comment by sidhu1012** [comment]
@oscarbenjamin I have added some new tests for remaining cases.Please review them, I think it isn't breaking any code and is playing nice.

**Comment by sidhu1012** [comment]
@oscarbenjamin If these changes are correct then shall I reflect them to entire codebase?

File: sympy/core/evalf.py [hunk 2]
```diff
@@ -1375,9 +1388,77 @@ def evalf(x, prec, options):
         check_target(x, r, prec)
     return r
 
+def _evalf_options(x, prec, options):
```
Comment (oscarbenjamin) [inline sympy/core/evalf.py:1391]: This function should just be called `evalf_options` i.e. no underscore because it's not private.

**Comment by oscarbenjamin** [comment]
The tests for all the different combinations are long and repetitive. It should be possible to combine them together and make them more systematic e.g.:
```
class A: ...
class B: ...
class C: ...

all_classes = [A, B, C]

def test_in_finiteset(cls): ...
def test_in_matrix(cls): ...
def test_in_sin(cls): ...

all_test_funcs = [test_in_finiteset, ...]

for cls in all_classes:
    for func in all_test_funcs:
        func(cls)
```

The most important cases to try and test for are the ones that users are likely to have used and that we want to protect for backwards compatibility. A reasonable guess at that can be looking at the ways that `EvalfMixin` is used around the codebase.

I'm still not really happy with "_eval_evalf_options" as a name but I haven't thought of anything better.

**Comment by sidhu1012** [comment]
> I'm still not really happy with "_eval_evalf_options" as a name but I haven't thought of anything better.


We can do `function overloading` and keep the name `_eval_evalf`.

It would be easier for downstream libraries too because they would need to change just the paramter and recursive call.

**Comment by oscarbenjamin** [comment]
> We can do `function overloading` and keep the name `_eval_evalf`.

If downstream code has defined an `_eval_evalf` method that only has a single parameter then calling it with extra arguments will give a TypeError. That means anyone who has made an `_eval_evalf` function would find that their code is broken by the change in sympy.

**Comment by sidhu1012** [comment]
@oscarbenjamin I have added the test cases , please review.

While adding test cases I realized that we should add `_eval_evalf_options` where ever necessary but shouldn't replace `_eval_evalf` yet for backward compatibility . So instead of replacing `_eval_evalf` , classes should have both the functions.

File: sympy/sets/sets.py [hunk 3]
```diff
@@ -1883,6 +1883,10 @@ def as_relational(self, symbol):
     def compare(self, other):
         return (hash(self) - hash(other))
 
+    def _eval_evalf_options(self, prec, options):
+        from sympy.core.evalf import evalf_options
+        return FiniteSet(*[evalf_options(elem, prec, options) for elem in self])
+
     def _eval_evalf(self, prec):
         return FiniteSet(*[elem.evalf(n=prec_to_dps(prec)) for elem in self])
```
Comment (oscarbenjamin) [inline sympy/sets/sets.py:1891]: One of these methods should be calling the other.

It probably makes sense to add a default implementation of `_eval_evalf` to `EvalfMixin` that calls `_eval_evalf_options`.

Comment (sidhu1012) [reply sympy/sets/sets.py:1891]: We can change implementation of `_eval_evalf` in Sympy codebase but changing default `_eval_evalf` won't help because they are overridden in every class.


We can't even call `_eval_evalf_options`  from `_eval_evalf` because we won't have `options` parameter.

Comment (oscarbenjamin) [reply sympy/sets/sets.py:1891]: It's helpful to add in the default implementation because it means that if downstream code is calling `_eval_evalf` directly on a sympy class then that will still work. It also means that we don't need to duplicate the methods like above.

Comment (sidhu1012) [reply sympy/sets/sets.py:1891]: Current changes in `.evalf()` would call `_eval_evalf_options` first , it fails then would try with `_eval_evalf` making both work.

I added back `_eval_evalf` because if some method class isn't updated accordingly it doesn't fail, to maintain backward compatibility .

We can't call `_eval_evalf_options` in any way from `_eval_evalf` because of missing parameter `options` and would give `TypeError`, whole purpose of `_eval_evalf_options` is to handle `options` because `_eval_evalf` can not.

Comment (oscarbenjamin) [reply sympy/sets/sets.py:1891]: The default `_eval_evalf` method in `EvalfMixin` could be made to call `_eval_evalf_options` passing the default values for the options.

Comment (sidhu1012) [reply sympy/sets/sets.py:1891]: If we change the default `_eval_evalf` method in `EvalfMixin` still we would need to duplicate the method because every every class calls `evalf` on its args differently in `_eval_evalf` that's why they need to be overridden in every class.

as in `FiniteSet` it is -
https://github.com/sympy/sympy/blob/d053c1d4c3ca5420851dcd9a30dfc0e7cee98a48/sympy/sets/sets.py#L1908

but in it `Interval` -
https://github.com/sympy/sympy/blob/d053c1d4c3ca5420851dcd9a30dfc0e7cee98a48/sympy/sets/sets.py#L1106

Comment (oscarbenjamin) [reply sympy/sets/sets.py:1891]: As long as EvalfMixin makes sure that both of those end up calling `_eval_evalf_options` then there should be no need to duplicate anything.

Comment (sidhu1012) [reply sympy/sets/sets.py:1891]: Yes, got it. Sorry mistook `_eval_evalf_options` with `evalf_options`.

**Comment by sidhu1012** [comment]
@oscarbenjamin I think all possible scenarios have been handled so shall I reflect changes everywhere?

**Comment by oscarbenjamin** [comment]
I'm not sure I understand the point of the different classes that are tested A, B and C. They seem unnecessarily complicated for what is tested. The most important case to test is missing which is a subclass of Expr that only implements `_eval_evalf`. That can be:
```
class ExprEvalf(Expr):
    def _eval_evalf(self, prec):
        return Rational(1, 3).evalf(prec)
```
You should add to these tests for any class whose `_eval_evalf` methods are changed (`FiniteSet` etc) especially if they change how they recurse.

More comments/notes are needed to explain why some methods are deprecated and why some call others etc. Especially since this is a change compared to before there needs to be a clear explanation of why it is changed.

Make sure to build the html docs and check to see what it looks like there: it is clear to users what they should do and what is deprecated and why etc? Separately you should make sure with comments in the code that it is clear to anyone reading the code why a deprecated method is being called.

There should be three examples in the `EvalfMixin` docstring: a Basic subclass, an Expr subclass and a non-Basic class. Each of them should illustrate the precise minimum that is needed to make evalf work for a class.

**Comment by oscarbenjamin** [comment]
Is it necessary to defined `is_number` or `is_Number`? One is used in the docstring and another in the tests.

File: sympy/core/tests/test_evalf.py [hunk 4]
```diff
@@ -579,3 +580,102 @@ def test_issue_13425():
...
+
+    class B(X, Expr):
+        pass
+
+    class C(Basic, X):
+        def could_extract_minus_sign(self):
+            return False
+
+        def as_coefficient(self, value):
+            return
```
Comment (oscarbenjamin) [inline sympy/core/tests/test_evalf.py:657]: Why are these methods needed? Are they used by `EvalfMixin`?

**Comment by oscarbenjamin** [comment]
> There should be three examples in the `EvalfMixin` docstring: a Basic subclass, an Expr subclass and a non-Basic class. Each of them should illustrate the precise minimum that is needed to make evalf work for a class.

A fourth example is also needed which shows how to write an Expr subclass that works both with old and new sympy versions.

It's important to be clear when making a change like this what has changed and why and how others should manage the change in their own code.

File: sympy/core/tests/test_evalf.py (hunk 4 above)
Comment (sidhu1012) [reply sympy/core/tests/test_evalf.py:657]: They were needed in case of sub class of `Basic` in case of `sin`.

**Comment by sidhu1012** [comment]
> Is it necessary to defined `is_number` or `is_Number`? One is used in the docstring and another in the tests.

Yes they are needed for independent classes inheriting `EvalfMixin`.  `is_number` or `is_Number` and `func(self)` is required.

**Comment by sidhu1012** [comment]
> A fourth example is also needed which shows how to write an Expr subclass that works both with old and new sympy versions.

I'm not sure how to explain this with an example, can you please suggest an example.

**Comment by sidhu1012** [comment]
> I'm not sure I understand the point of the different classes that are tested A, B and C. They seem unnecessarily complicated for what is tested.

Those tests are just checking that `_eval_evalf_options` is working fine for various possible classes.

**Comment by oscarbenjamin** [comment]
> > A fourth example is also needed which shows how to write an Expr subclass that works both with old and new sympy versions.
> 
> I'm not sure how to explain this with an example, can you please suggest an example.

If a user updates their `_eval_evalf` method to `_eval_evalf_options` then their code will work with the changed version of sympy. It will still be possible to call `_eval_evalf` in the changed version of sympy because `EvalfMixin` provides that. However if they also want to provide `_eval_evalf` to work with older sympy versions then they will need to copy the `_eval_evalf` implementation from here. They will also need a replacement for `evalf_options`. We need to think about how to do this to make it as easy as possible for them and then document that.

**Comment by sidhu1012** [comment]
> If a user updates their `_eval_evalf` method to `_eval_evalf_options` then their code will work with the changed version of sympy. It will still be possible to call `_eval_evalf` in the changed version of sympy because `EvalfMixin` provides that. However if they also want to provide `_eval_evalf` to work with older sympy versions then they will need to copy the `_eval_evalf` implementation from here. They will also need a replacement for `evalf_options`. We need to think about how to do this to make it as easy as possible for them and then document that.


The best way for `.evalf` work with older sympy versions is to implement `_eval_evalf` and recurse using `.evalf()`.
I'll update the docstring!

**Comment by sidhu1012** [comment]
> Make sure to build the html docs and check to see what it looks like there: it is clear to users what they should do and what is deprecated and why etc? 

Where should I add html doc and what should be it's content?

**Comment by sidhu1012** [comment]
@oscarbenjamin I have updated the test cases as well as docstrings. Please go through them once and review them.

**Comment by sidhu1012** [comment]
> Make sure to build the html docs and check to see what it looks like there: it is clear to users what they should do and what is deprecated and why etc?

@oscarbenjamin Can you please help me out with it. I'm not sure what should I do for this.

**Comment by sidhu1012** [comment]
@czgdp1807 A couple of stochastic process tests are failing I am not sure why as I didn't do any stats related change. Can you have a look once. I'm trying on my end.

**Comment by czgdp1807** [comment]
```python
executable:         /usr/bin/python3  (3.6.9-final-0) [CPython]
architecture:       64-bit
cache:              yes
ground types:       python 
numpy:              1.19.2
hash randomization: on (PYTHONHASHSEED=790239741)

sympy/stats/stochastic_process_types.py[11] ...........                      [OK]
```
On latest `master` the doctests are passing. So probably, the changes are affecting the results generated by other modules.

**Comment by oscarbenjamin** [comment]
The tests that are run on github are not run directly on the code here. Rather a branch is created from master and this PR is merged into that branch and then the tests are run on that.

If you are seeing failures that you can't reproduce locally you might need to merge in master to see them.

**Comment by sidhu1012** [comment]
> > Make sure to build the html docs and check to see what it looks like there: it is clear to users what they should do and what is deprecated and why etc? 
> 
> Where should I add html doc and what should be it's content?
> 
> 

@oscarbenjamin Can you please help me out with it, as I'm not sure how to do this.

**Comment by oscarbenjamin** [comment]
The html docs are built form the rst files in the `doc` folder. They also include the content of the docstrings. I'm suggesting that you should build the docs, open the html in a browser and read it to see if someone looking at the docs would understand how to implement `.evalf` correctly.

**Comment by sidhu1012** [comment]
@oscarbenjamin I'm not being able to generate `html docs`. I have been trying to generate them for sometime, but everytime it fails to generate them. I'm using wsl2 on windows 10.

**Comment by oscarbenjamin** [comment]
How exactly does it fail?

**Comment by sidhu1012** [comment]
1) imagemagick docbook2x graphviz
imagemagick: command not found

2) make html
rm -rf _build/logo
mkdir -p _build/logo
python ./generate_logos.py -d
make: python: Command not found
make: *** [Makefile:144: logo] Error 127

3) firefox _build/html/index.html
Unable to init server: Broadway display type not supported: yourmachine.yourdomain.com:0.0
Error: cannot open display: yourmachine.yourdomain.com:0.0

**Comment by sidhu1012** [comment]
@oscarbenjamin While working , I just found that `_eval_evalf` shouldn't be removed completely because it can be pretty useful in cases where we don't need `evalf()` but need mpmath functions , since they don't bother about `options` parameter.

**Comment by czgdp1807** [comment]
The code quality tests on this PR are failing.

**Comment by sidhu1012** [comment]
> The code quality tests on this PR are failing.

Yes I'll fix them soon, a bit busy these days.

**Comment by sidhu1012** [comment]
Feel free to take it over as I may not get much to complete it in upcoming months.
